src/util/oauth_utils.js

```diff
@@ -0,0 +1,134 @@
+'use strict';
+
+const querystring = require('querystring');
+const config = require('../../config');
+const { make_http_request } = require('./http_utils');
+const { read_stream_join } = require('./buffer_utils');
+
+function _default_error_factory(message) {
+    return new Error(message);
+}
+
+async function trade_grant_code_for_access_token(
+    oauth_service,
+    client_id,
+    client_secret,
+    redirect_host,
+    grant_code,
+    make_error = _default_error_factory
+) {
+    const token_endpoint = new URL(config.OAUTH_TOKEN_ENDPOINT, `https://${oauth_service}`);
+    const redirect_uri = new URL(config.OAUTH_REDIRECT_ENDPOINT, redirect_host);
+    let response;
+
+    try {
+        response = await make_http_request(
+            {
+                method: 'POST',
+                port: token_endpoint.port || '443',
+                hostname: token_endpoint.hostname,
+                path: token_endpoint.pathname,
+                rejectUnauthorized: false,
+                headers: {
+                    'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8',
+                    'X-CSRF-Token': 'not-empty'
+                }
+            },
+            querystring.stringify({
+                grant_type: 'authorization_code',
+                code: grant_code,
+                client_id: client_id,
+                client_secret: client_secret,
+                redirect_uri: redirect_uri.toString()
+            }),
+            'utf8'
+        );
+    } catch (err) {
+        throw make_error('OpenShift api endpoint does not response');
+    }
+
+    const status_code = response.statusCode;
+    const buffer = await read_stream_join(response);
+    const body = buffer.toString('utf8');
+
+    if (status_code !== 200) {
+        throw make_error(`could not acquire an oauth access token, response status code was ${
+            status_code
+        } with a body of ${
+            body
+        }`);
+    }
+
+    try {
+        const token_info = JSON.parse(body);
+        const required_scope = config.OAUTH_REQUIRED_SCOPE;
+        if (token_info.scope !== required_scope) {
+             throw make_error(`user did not grant access for "${required_scope}" scope`);
+        }
+        if (token_info.token_type !== 'Bearer') {
+            throw make_error('access token is not a of type "Bearer"');
+        }
+        return token_info;
+
+    } catch (err) {
+        throw make_error('mailformed access token json response');
+    }
+}
+
+async function review_token(
+    api_hostname,
+    api_port = '443',
+    sa_token,
+    oauth_access_token,
+    make_error = _default_error_factory
+) {
+    let response;
+    try {
+        response = await make_http_request(
+            {
+                method: 'POST',
+                hostname: api_hostname,
+                port: api_port,
+                path: config.KUBE_API_ENDPOINTS.TOKEN_REVIEW,
+                rejectUnauthorized: false,
```

Comment:
What does this line do.

Reply:
@enj It tell's our http client to accept a self signed certificate from the server side.
We have to allow it for any test system to work.
However for secure mode we will replace it with a `config.REJECT_UNAUTHORIZED_HTTPS_CERTIFICATE = true`.

Reply:
> @enj It tell's our http client to accept a self signed certificate from the server side.
> We have to allow it for any test system to work.
> However for secure mode we will replace it with a `config.REJECT_UNAUTHORIZED_HTTPS_CERTIFICATE = true`.

Since this code will only work when you have a kube cluster, why not always use `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt` to verify the connection?

Reply:
Thanks, opened an issues to track this across the code - https://github.com/noobaa/noobaa-core/issues/5578